presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractMapBlock.java

```diff
@@ -335,10 +341,58 @@ public boolean isNull(int position)
         return mapIsNull != null && mapIsNull[position + getOffsetBase()];
     }
 
+    public boolean areHashTablesCreated()
+    {
+        return getHashTables().get().isPresent();
+    }
+
     private void checkReadablePosition(int position)
     {
         if (position < 0 || position >= getPositionCount()) {
             throw new IllegalArgumentException("position is not valid");
         }
     }
+
+    public static class HashTables
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(HashTables.class).instanceSize();
+
+        // Hash to location in map. Writes to the field by MapBlock is protected by "HashTables" monitor in MapBlock.
+        // MapBlockBuilder would always have its own non-null hashTables and do not need to be synchronized.
```

Comment:
`MapBlockBuilder instances have their dedicated hashTables instances, so the write accesses to the hashTables fields do not need to be synchronized in that class.`